core/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesShardsStoresAction.java

```diff
@@ -0,0 +1,220 @@
+                break;
+            case ALL:
+                shardRoutings = indexRoutingTables.allShards();
+                break;
+            default:
+                // bogus
+                listener.onFailure(new IllegalArgumentException("Invalid shard state criteria"));
+                return;
+        }
+        final Set<String> requestedIndices = new HashSet<>(Arrays.asList(request.indices()));
+        final Set<ShardId> shardsToFetch = new HashSet<>();
+        for (ShardRouting shard : shardRoutings) {
+            if (requestedIndices.size() == 0 || requestedIndices.contains(shard.shardId().index().getName())) {
+                // Should we also filter by shard.primary(),
+                // in doing so, no store info will be fetched
+                // for UNASSIGNED replicas (yellow state)
+                shardsToFetch.add(shard.shardId());
+            }
+        }
+
+        // async fetch store infos for relevant shard from all the nodes
+        if (shardsToFetch.size() > 0) {
+            final DiscoveryNodes nodes = state.nodes();
+            final MetaData metaData = state.metaData();
+            // NOTE: instead of fetching shard store info one by one from every node (nShards * nNodes requests)
+            // we could fetch all shard store info from every node once (nNodes requests)
+            // we have to implement a TransportNodesAction instead of using TransportNodesListGatewayStartedShards
+            // for fetching shard stores info, that operates on a list of shards instead of a single shard
+            AsyncShardsStoresInfoFetches asyncShardsStoresInfoFetches = new AsyncShardsStoresInfoFetches(nodes, metaData, shardsToFetch.size(), listener);
+            asyncShardsStoresInfoFetches.fetch(shardsToFetch);
+        } else {
+            listener.onResponse(new IndicesShardsStoresResponse());
+        }
+    }
+
+    @Override
+    protected ClusterBlockException checkBlock(IndicesShardsStoresRequest request, ClusterState state) {
+        return state.blocks().indicesBlockedException(ClusterBlockLevel.METADATA_READ, request.indices());
+    }
+
+    private class AsyncShardsStoresInfoFetches {
+        private final DiscoveryNodes nodes;
+        private final MetaData metaData;
+        private final CountDown expectedOps;
+        private final Queue<InternalAsyncFetch.Response> fetchResponses;
+        private final ActionListener<IndicesShardsStoresResponse> listener;
+
+        AsyncShardsStoresInfoFetches(DiscoveryNodes nodes, MetaData metaData, int expectedOps, ActionListener<IndicesShardsStoresResponse> listener) {
+            this.nodes = nodes;
+            this.metaData = metaData;
+            this.listener = listener;
+            this.expectedOps = new CountDown(expectedOps);
+            this.fetchResponses = new ConcurrentLinkedQueue<>();
+        }
+
+        void fetch(Collection<ShardId> shardIds) {
```

Comment:
can we make sure that expectedOps is the size of shardIds by moving the shardIds to the constructor and capturing both?